grafana/client.go

```diff
@@ -0,0 +1,292 @@
+	for _, a := range alerts {
+		fmt.Fprintf(w,
+			"%s\t%s\t%s\t%d\t%s\n",
+			a.Time.UTC().Format(time.RFC3339), a.PrevState, a.NewState, a.ID, a.Text,
+		)
+	}
+
+	// Flush writes to the underlying writer
+	w.Flush()
+
+	return b.String()
+}
+
+type AnnotationsQueryParams struct {
+	Limit        *int
+	AlertID      *int
+	DashboardID  *int
+	DashboardUID *string // when dashboardUID presents, dashboardId would be ignored by /annotations API
+	Type         *string
+	From         *time.Time
+	To           *time.Time
+}
+
+type PostAnnotation struct {
+	DashboardID  *int
+	PanelID      *int
+	DashboardUID string
+	Time         *time.Time
+	TimeEnd      *time.Time
+	Tags         []string
+	Text         string
+}
+
+func (g *Client) GetAnnotations(params AnnotationsQueryParams) ([]Annotation, *resty.Response, error) {
+	query := make(url.Values)
+	if params.Limit != nil {
+		query.Set("limit", fmt.Sprintf("%d", *params.Limit))
+	}
+	if params.AlertID != nil {
+		query.Set("alertId", fmt.Sprintf("%d", *params.AlertID))
+	}
+	if params.DashboardID != nil {
+		query.Set("dashboardId", fmt.Sprintf("%d", *params.DashboardID))
+	}
+	if params.DashboardUID != nil {
+		query.Set("dashboardUID", *params.DashboardUID)
+	}
+	// type can be: alert or annotation
+	if params.Type != nil {
+		query.Set("type", *params.Type)
+	}
+
+	// Grafana issue https://github.com/grafana/grafana/issues/63130
+	if (params.From != nil && params.To == nil) || (params.To != nil && params.From == nil) {
+		return nil, nil, fmt.Errorf("both From and To must be set")
+	}
+
+	if params.From != nil {
+		query.Set("from", fmt.Sprintf("%d", params.From.UnixMilli()))
+	}
+	if params.To != nil {
+		query.Set("to", fmt.Sprintf("%d", params.To.UnixMilli()))
+	}
+
+	var result []Annotation
+	r, err := g.resty.R().
+		SetResult(&result).
+		SetQueryString(query.Encode()).
+		Get("/api/annotations")
+	return result, r, err
+}
+
+func (g *Client) PostAnnotation(annotation PostAnnotation) (*resty.Response, error) {
+	a := map[string]interface{}{
+		"dashboardUID": annotation.DashboardUID,
+		"tags":         annotation.Tags,
+		"text":         annotation.Text,
+	}
+	if annotation.DashboardID != nil {
+		a["dashboardId"] = *annotation.DashboardID
+	}
+	if annotation.PanelID != nil {
+		a["panelId"] = *annotation.PanelID
+	}
+	if annotation.Time != nil {
+		a["time"] = annotation.Time.UnixMilli()
+	}
+	if annotation.TimeEnd != nil {
+		a["timeEnd"] = annotation.TimeEnd.UnixMilli()
+	}
+	return g.resty.R().
+		SetBody(a).
+		Post("/api/annotations")
+}
+
+// ruler API is deprecated https://github.com/grafana/grafana/issues/74434
+func (g *Client) GetAlertsForDashboard(dashboardUID string) (map[string][]interface{}, *resty.Response, error) {
+	var result map[string][]interface{}
+	r, err := g.resty.R().SetResult(&result).Get("/api/ruler/grafana/api/v1/rules?dashboard_uid=" + url.QueryEscape(dashboardUID))
+	return result, r, err
+}
+
+type AlertQuery interface{}
```

Comment:
Do we plan to add more on top of it? Empty type.

Reply:
@skudasov I removed this type for now